scripts/run_mesh_experiments.py

```diff
@@ -0,0 +1,293 @@
+    dense_size = params["dense_size"]
+    attention = params["attention"]
+    architecture = params["architecture"]
+    epochs = params["epochs"]
+    vocabulary_size = params["vocabulary_size"]
+    sequence_length = params["sequence_length"]
+
+    logdir = "logs/scalars/" + datetime.now().strftime("%Y%m%d-%H%M%S") + f"-{learning_rate}-{batch_size}"
+    tensorboard = tf.keras.callbacks.TensorBoard(log_dir=logdir)
+
+    early_stopping = tf.keras.callbacks.EarlyStopping(patience=5, restore_best_weights=True)
+
+    def yield_data(X, Y, batch_size, shuffle=True):
+        while True:
+            if shuffle:
+                randomize = np.arange(len(X))
+                np.random.shuffle(randomize)
+                X = X[randomize]
+                Y = Y[randomize]
+            for i in range(0, X.shape[0], batch_size):
+                yield X[i:i+batch_size, :], Y[i:i+batch_size, :].todense()
+
+    train_data = yield_data(X_train, Y_train, batch_size)
+    test_data = yield_data(X_test, Y_test, batch_size)
+    steps_per_epoch = math.ceil(X_train.shape[0]/batch_size)
+    validation_steps = math.ceil(X_test.shape[0]/batch_size)
+
+    if architecture == 'cnn':
+        build_model_f = build_cnn_model
+    elif architecture == 'bilstm':
+        build_model_f = build_bilstm_model
+    else:
+        build_model_f = build_custom_bilstm_model
+    model = build_model_f(
+        learning_rate=learning_rate, nb_tags=nb_tags, attention=attention,
+        l2=l2, dropout=dropout, hidden_size=hidden_size, dense_size=dense_size,
+        vocabulary_size=vocabulary_size, sequence_length=sequence_length)
+    history = model.fit(x=train_data, steps_per_epoch=steps_per_epoch,
+                        validation_data=test_data, validation_steps=validation_steps,
+                        epochs=epochs, callbacks=[tensorboard, early_stopping])
+
+    steps = len(history.history["loss"]) * steps_per_epoch * batch_size
+
+    # evaluate                                                                                                                 
+    def predict(model, data_gen, steps):
+        Y_pred = []
+        for _ in tqdm(range(steps)):
+            X_batch, _ = next(data_gen)
+            Y_pred_batch = model(X_batch) > 0.5
+            Y_pred.append(csr_matrix(Y_pred_batch))
+        Y_pred = vstack(Y_pred)
+        return Y_pred
+ 
+    test_data = yield_data(X_test, Y_test, batch_size, shuffle=False)
+
+    Y_pred_test = predict(model, test_data, validation_steps)
+    f1_test = f1_score(Y_test, Y_pred_test, average='micro')
+    return f1_test, steps
+
+def experiment(data_path, params):
+    nb_tags = params["nb_tags"]
+    nb_examples_per_tag = params["nb_examples_per_tag"]
+    vocabulary_size = params["vocabulary_size"]
+    sequence_length = params["sequence_length"]
+    texts, tags = create_dataset(
+        data_path, nb_tags, nb_examples_per_tag)
+
+    nb_train = len(texts) - min(int(0.2*len(texts)), 10_000)
```

Comment:
So you are saying that the test set will have a maximum of 10000 docs? Why is that?

Reply:
Well I should have done this earlier tbh. Andrew Ng suggests that after a dataset size you do not need the typical 80/20 split but more of a 99/1. You really only need enough data to make a statistically robust estimate of the error or accuracy. Check https://www.coursera.org/lecture/machine-learning-projects/size-of-the-dev-and-test-sets-HOby4

Reply:
Yeah, I remember that from the course. The one thing to bear in mind though is that you have an extreme mutilabel problem, so a statistically robust sample might require more data. (maybe 10.000 is fine)

Reply:
It is a good point. i am actually using the same number used by deep mesh https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4908368/pdf/btw294.pdf
```
We sorted all downloaded citations by time. The latest 10 000 citations were used as validation set for tuning parameters of binary classifier
```
But keen to hear if there is a way or rule of thumb to decide how many examples are enough.